app/apollo/resolvers/group.js

```diff
@@ -622,46 +699,64 @@ const groupResolvers = {
       const user = whoIs(me);
 
       try {
-        logger.info({ req_id, user, org_id, uuid, clusters }, `${queryName} validating`);
+        logger.info({req_id, user, org_id, uuid, clusters}, `${queryName} validating`);
 
         validateString( 'org_id', org_id );
         validateString( 'uuid', uuid );
         clusters.forEach( value => validateString( 'clusters', value ) );
 
         // validate the group exits in the db first.
         const group = await models.Group.findOne({ org_id: org_id, uuid });
-        if(!group){
-          throw new NotFoundError(context.req.t('group uuid "{{uuid}}" not found', {'uuid':uuid}));
-        }
+        logger.info({req_id, user, org_id, uuid, clusters}, `${queryName} validating - found: ${!!group}`);
 
-        await validAuth(me, org_id, ACTIONS.MANAGE, TYPES.GROUP, queryName, context, [group.uuid, group.name]);
+        const identifiers = group ? [uuid, group.name] : [uuid];
+        await validAuth(me, org_id, ACTIONS.MANAGE, TYPES.GROUP, queryName, context, identifiers);
+        logger.info({req_id, user, org_id, uuid, clusters}, `${queryName} validating - authorized`);
 
-        logger.info({ req_id, user, org_id, uuid, clusters }, `${queryName} saving`);
+        if (!group) {
+          throw new NotFoundError(context.req.t('group uuid "{{uuid}}" not found', {'uuid':uuid}), context);
+        }
 
-        // update clusters group array with the above group
-        const res = await models.Cluster.updateMany(
-          {org_id: org_id, cluster_id: {$in: clusters}, 'groups.uuid': {$in: [uuid]}},
-          {$pull: {groups: {uuid}}});
+        let allAllowed = false;
+        try {
+          await validAuth(me, org_id, ACTIONS.READ, TYPES.CLUSTER, queryName, context);
+          allAllowed = true;
+        }
+        catch(e){ // If exception thrown, user does NOT have auth to all resources of this type, and code must later filter based on fine grained auth
+        }
+        logger.info({req_id, user, org_id, uuid, clusters}, `${queryName} validating - allAllowed: ${allAllowed}`);
 
-        pubSub.channelSubChangedFunc({org_id: org_id}, context);
+        let foundClusters = await commonClusterSearch(models, {org_id}, { limit: 0, skip: 0, startingAfter: null });
+
+        if (!allAllowed){
+          foundClusters = await filterResourcesToAllowed(me, org_id, ACTIONS.READ, TYPES.CLUSTER, foundClusters, context);
+          logger.info({req_id, user, org_id, uuid, clusters}, `${queryName} filtered resources to allowed`);
+        }
```

Comment:
Should throw a NFE if foundClusters length is not equal to passed in clusters length.

Reply:
Opinion: For ungroupClusters, a code comment and unit test would suffice as there are situations where we would want this function to continue forward in the situation where clusters may not exist but the user does have permission to use them.

Reply:
I agree, but like for `unassignClusterGroups` I think the situation we would want to continue is if a user tries to ungroup a cluster that does not exist, regardless of whether the user has permission to them or not.

Reply:
Resolved